releases/ReleaseCompatibility.md

```diff
@@ -0,0 +1,29 @@
+# Releases & Compatibility
+
+To date, we have been somewhat cavalier about making breaking changes in Fairlearn.
+Although we are currently pre-`v1.0.0` and hence without particular commitments to compatibility (as generally understood - see e.g. the [Semantic Versioning Scheme](https://semver.org/)), we should work to reduce the number of breaking changes we make.
+
+## The Problem
+
+At the time of writing, we are about to release `v0.4.6`, which has some substantial breaking changes in the Metrics area of Fairlearn.
+Before it, `v0.4.5` reworked a smaller subset of this functionality, which caused smaller breakages.
+There are also further changes to Metrics planned, which may well cause further breaks (although these should be more minor).
+All this is obviously undesirable from a user standpoint - these look like 'patch' level releases, but are actually breaking their code.
+Concretely, we've had users install Fairlearn with `pip`, and then find themselves unable to run Notebooks from our GitHub project - not because the functionality was missing, but because it had been renamed.
+
+We do not need deprecation policies as elaborate of those of [SciKit-Learn](https://numpy.org/neps/nep-0023-backwards-compatibility.html) or [NumPy](https://numpy.org/neps/nep-0023-backwards-compatibility.html) - indeed, policies such as those would be ridiculous given the size of our code and user bases.
+However, we do need to start to move in that direction, or we will not be able to grow our user base due to chaos in the code.
+
+## Proposed Solution
+
+Starting with `v0.5.0` we should make a commitment that anything which works at `v0.n.m_0` will also work for `v0.n.m` so long as `m >= m_0`.
+
+We will enforce this using builds which run the notebooks currently in the repository against versions of Fairlearn published on PyPI.
+We already have a build which checks that our notebooks can run against the latest published version of Fairlearn (due to the metrics changes, this build is a steady red); this would be an extension of that functionality.
+For full compliance with the above, we will have to devise a means of allowing each notebook to specify its `m_0`.
+Our notebooks do not provide comprehensive coverage of Fairlearn, but enforcing a constraint like this will be a dramatic improvement over the current situation (particularly since new users will likely try our notebooks first).
+
+In order to support less mature functionality, we should also add a `fairlearn.experimental` package (see [a similar namespace in SciKit-Learn](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.experimental)).
```

Comment:
I'm not sure we need this yet. I guess I wonder how much overhead this would be on the testing side and what we might put into this subpackage.

Reply:
I'm not opposed to `fairlearn.experimental` in general, although I don't think it would have solved any of our issues. We're reworking all the core parts and that's why we're causing issues right now. 

Scenarios for `fairlearn.experimental` in my view are:
- new mitigation techniques that we're trying out (e.g. for research purposes) and that we don't necessarily want users to actively use unless they're aware of the caveats
- let's say someone wants to replicate our charts in matplotlib, that may be very incomplete at first and may need some work to get in a state where there's enough functionality (but would be too much for a single PR)

This is specific to Python code, so this wouldn't be usable for the dashboard and adding new components, right?

Reply:
Yes, `fairlearn.experimental` certainly wouldn't have helped with our current refactorings. The idea is that going forward, it gives contributors somewhere to 'play' without getting stuck supporting their earlier poor choices (which are inevitable, since users always come up with ingenious ways of breaking things :-) ).

For the UX, I don't know remotely enough about TS to comment. Perhaps something similar could happen there.

I'd very much appreciate experiences from other open source projects here.

Reply:
TS does have namespaces, so I've added a note.